### PR TITLE
docs: explain json output of status

### PR DESCRIPTION
**Pull request by orndorffgrant**
no-lp no-jira

Fixes: #2178

## Checklist

 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [x] I have updated or added any documentation accordingly

## Does this PR require extra reviews?

 - [ ] Yes
 - [x] No

### CODE REVIEW

File: docs/explanations/status_columns.md [hunk 1]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
 It's important to keep in mind that any feature defined like this will be
 listed, even if it is invalid or typed the wrong way. Those appear in `status`
 output for informational and debugging purposes.
+
+## Machine Readable Output
```
Comment (s-makin) [inline docs/explanations/status_columns.md:105]:
```suggestion
## Machine-readable output
```

File: docs/explanations/status_columns.md [hunk 2]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+
+The `pro status` command supports a `--format` flag with options including `json` and `yaml`. These result in a machine-readable form of the information presented in the `pro status` command.
```
Comment (s-makin) [inline docs/explanations/status_columns.md:107]:
```suggestion
The `pro status` command supports a `--format` flag with options including `json` and `yaml`. These result in a machine-readable form of the information presented by the `pro status` command.
```

File: docs/explanations/status_columns.md [hunk 3]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+
+```{note}
+`pro status` should return the same results with and without sudo/root, but earlier versions did not always and we would still recommend using sudo if possible.
```
Comment (s-makin) [inline docs/explanations/status_columns.md:110]:
```suggestion
`pro status` should return the same results whether using `sudo` or not, but earlier versions did not always do this. We recommend using `sudo` whenever possible.
```

File: docs/explanations/status_columns.md [hunk 4]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+    ...otherServiceStatusObjects
+  ],
+  "simulated": false,
+  "version": "27.13.6~18.04.1",
+  "warnings": []
+}
+```
+
+Some particularly important attributes in the output include:
+* `attached`: This boolean value indicates whether this machine is attached to an Ubuntu Pro account or not. Note that this does not imply that any particular service (e.g. esm-infra) is enabled. You must check the individual service object for that status.
```
Comment (s-makin) [inline docs/explanations/status_columns.md:183]:
```suggestion
* `attached`: This boolean value indicates whether or not this machine is attached to an Ubuntu Pro account. This does not tell you if any particular service (e.g. `esm-infra`) is enabled. You must check the individual service object for that status.
```

Comment (s-makin) [reply docs/explanations/status_columns.md:183]: As a quick note, I'm not sure about the wording for the final sentence - I've not heard of a "service object" before so I'm not sure what the sentence means. Is there a clearer way we can express this?

File: docs/explanations/status_columns.md [hunk 5]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+* `expires`: This is the date that the Ubuntu Pro subscription is valid until.
+* `services`: This is a list of Ubuntu Pro services. Each item has its own attributes. Widely applicable services include those with `name` equal to `esm-infra`, `esm-apps`, and `livepatch`. Some important fields in each service object are:
+  * `name`: The name of the service.
+  * `entitled`: A boolean indicating whether the attached Ubuntu Pro account is allowed to enable this service.
+  * `status`: An enum value indicating the services current status on the machine. Possible values are `enabled`, `disabled`, `n/a`, and `warning`. Any value other than `enabled` should be treated as if the service is not enabled and not working properly on the machine.
```
Comment (s-makin) [inline docs/explanations/status_columns.md:188]:
```suggestion
  * `status`: An enum value indicating the service's current status on the machine. Possible values are `enabled`, `disabled`, `n/a`, and `warning`. Any value other than `enabled` should be treated as if the service is not enabled and not working properly on the machine.
```

File: docs/explanations/status_columns.md [hunk 6]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+
+For example, if you want to programatically determine if esm-infra is enabled and working on a particular machine, you can use the following command:
```
Comment (s-makin) [inline docs/explanations/status_columns.md:190]:
```suggestion
For example, if you want to programatically determine if `esm-infra` is enabled and working on a particular machine, you can use the following command:
```

Comment (s-makin) [inline docs/explanations/status_columns.md:190]:
```suggestion
For example, if you want to determine if esm-infra is enabled and working on a particular machine, you can use the following command:
```

File: docs/explanations/status_columns.md [hunk 7]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
...
+```shell
+sudo pro status --format=json | jq '.services[] | select(.name == "esm-infra").status == "enabled"'
+```
+That command will print `true` if esm-infra is enabled, or `false` if it is not.
+
```
Comment (s-makin) [inline docs/explanations/status_columns.md:195]: Side question - what should they do if it's disabled and they expected it to be enabled and working?

File: docs/explanations/status_columns.md (hunk 4 above)
Comment (orndorffgrant) [reply docs/explanations/status_columns.md]: I reworded it - let me know if it makes more sense now.

File: docs/explanations/status_columns.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply docs/explanations/status_columns.md]: It's jargon for "there's a finite list of enumerated values that could be here". I changed it to be simpler

File: docs/explanations/status_columns.md (hunk 7 above)
Comment (orndorffgrant) [reply docs/explanations/status_columns.md:195]: It depends on what the value is.
If it's "disabled", then run `pro enable`.
If it's `n/a` then you don't have the option to enable it.
If it's `warning` then something specific is wrong and you need to go read the warning detailed message to figure out what to do.
I'm not sure if all of that belongs here or not - I could be convinced either way - what do you think?

File: docs/explanations/status_columns.md [hunk 8]
```diff
@@ -101,3 +101,98 @@ allow_beta: True
 It's important to keep in mind that any feature defined like this will be
 listed, even if it is invalid or typed the wrong way. Those appear in `status`
 output for informational and debugging purposes.
+
+## Machine-readable Output
```
Comment (s-makin) [inline docs/explanations/status_columns.md:105]:
```suggestion
## Machine-readable output
```

File: docs/explanations/status_columns.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (s-makin) [reply docs/explanations/status_columns.md]: Much clearer!

File: docs/explanations/status_columns.md (hunk 7 above)
Comment (s-makin) [reply docs/explanations/status_columns.md:195]: I might have misunderstood, but the command reads to me like you're only checking if it's enabled (or not) - does the output actually tell you more than that whether it's available-but-disabled, n/a, or broken? So my initial question was more along the lines of "what would they need to follow up with if they get "false" as the result?"

File: docs/explanations/status_columns.md (hunk 8 above)
Comment (orndorffgrant) [reply docs/explanations/status_columns.md]: fixed for real this time :)

File: docs/explanations/status_columns.md (hunk 7 above)
Comment (orndorffgrant) [reply docs/explanations/status_columns.md:202]: I think I see your concern now. I've modified the example and the preceding text a bit. LMK what you think

File: docs/explanations/status_columns.md [hunk 9]
```diff
@@ -101,3 +101,105 @@ allow_beta: True
...
+  ],
+  "simulated": false,
+  "version": "27.13.6~18.04.1",
+  "warnings": []
+}
+```
+
+Some particularly important attributes in the output include:
+* `attached`: This boolean value indicates whether this machine is attached to an Ubuntu Pro account. This does not tell you if any particular service (e.g. `esm-infra`) is enabled. You must check the individual service item in the `services` list for that status (described below).
+* `expires`: This is the date that the Ubuntu Pro subscription is valid until (in RFC3339 format). After this date has passed the machine should be treated as if not attached and no services are enabled. `attached` may still say `true` and services may still say they are `entitled` and `enabled`, but if the `expires` date has passed, you should assume the services are not functioning.
```
[the remark this answers is not in the document]
Comment (s-makin) [reply docs/explanations/status_columns.md:187]: That's a strange behaviour -- does it not warn the user that the services are inactive when the expiry date is exceeded? Or at least prompt them to resubscribe?

(The amended wording is fine btw, I'm just trying to understand why it does that)

File: docs/explanations/status_columns.md (hunk 7 above)
Comment (s-makin) [reply docs/explanations/status_columns.md:202]: Makes much more sense! Thanks for changing it :)

**Review by s-makin — APPROVED** [review]
LGTM. My final question isn't about the contents of the PR, we can talk about that separately.

File: docs/explanations/status_columns.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply docs/explanations/status_columns.md:187]: Elsewhere (in APT and MOTD) yes, but not in the output of this command. This is a known issue but I don't think we had something to track it before so just created #2553

**Pull request merged**